### PR TITLE
add pruning when starting a new consumer

### PR DESCRIPTION
**Pull request by preciselytord**

### CODE REVIEW

File: cmd/cleanup/cleanup.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	if len(os.Args) > 2 {
+		group := os.Args[1]
+		stream := os.Args[2]
+		// prune old inactive consumers
+		cu := redisClient.XInfoConsumers(context.Background(), stream, group)
+		cs, err := cu.Result()
+		if err != nil {
+			logrus.WithError(err).Fatalf("failed to list consumers for stream: %q group: %q", stream, group)
+		}
+		fmt.Printf("Found %d consumers \n", len(cs))
```
Comment (ndreas) [inline cmd/cleanup/cleanup.go]: Use logrus instead, since it's used above?

File: cmd/cleanup/cleanup.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		for x, consumer := range cs {
+			fmt.Printf("%d: Checking consumer: %q : ", x, consumer.Name)
+			if consumer.Pending == 0 && // don't remove with pending msg
+				consumer.Idle > time.Hour.Milliseconds()*24 { // respect PruneTimeout before trying to delete
+
+				fmt.Print("removing\n")
```
Comment (ndreas) [inline cmd/cleanup/cleanup.go]: Logrus or `fmt.Println()`

File: cmd/cleanup/cleanup.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		cu := redisClient.XInfoConsumers(context.Background(), stream, group)
+		cs, err := cu.Result()
+		if err != nil {
+			logrus.WithError(err).Fatalf("failed to list consumers for stream: %q group: %q", stream, group)
+		}
+		fmt.Printf("Found %d consumers \n", len(cs))
+		for x, consumer := range cs {
+			fmt.Printf("%d: Checking consumer: %q : ", x, consumer.Name)
+			if consumer.Pending == 0 && // don't remove with pending msg
+				consumer.Idle > time.Hour.Milliseconds()*24 { // respect PruneTimeout before trying to delete
```
Comment (ndreas) [inline cmd/cleanup/cleanup.go]: Is this related to `PruneTimeout` in `consumer.go`? If not, why 24 milliseconds?

File: consumer.go [hunk 4]
```diff
@@ -218,9 +223,57 @@ func (c *Consumer) Run() {
...
+			continue
+		}
+		for _, consumer := range cs {
+			if consumer.Name != c.options.Name && // don't remove self
+				consumer.Pending == 0 && // don't remove with pending msg
+				consumer.Idle > c.options.PruneTimeout.Milliseconds() { // respect PruneTimeout before trying to delete
+				rCmd := c.redis.XGroupDelConsumer(ctx, stream, c.options.GroupName, consumer.Name)
+				pendingMsgCount, err := rCmd.Result()
+				if err != nil {
+					c.Errors <- errors.New(fmt.Sprintf("failed to remove consumer: %q for stream: %q group: %q", consumer.Name, stream, c.options.GroupName))
```
Comment (ndreas) [inline consumer.go]: moar `fmt.Errorf()`

File: consumer.go [hunk 5]
```diff
@@ -218,9 +223,57 @@ func (c *Consumer) Run() {
...
+				}
+				if pendingMsgCount > 0 {
+					c.Errors <- errors.New(fmt.Sprintf("remove consumer: %q with pending messages: %d for stream: %q group: %q", consumer.Name, pendingMsgCount, stream, c.options.GroupName))
```
Comment (ndreas) [inline consumer.go]: Another `fmt.Errorf()` opportunity.

File: consumer.go [hunk 6]
```diff
@@ -218,9 +223,57 @@ func (c *Consumer) Run() {
...
+	if err != nil {
+		c.Errors <- errors.New("failed to list streams from redis")
+	}
+
+	// prune old inactive consumers
+	for _, stream := range streams {
+		cu := c.redis.XInfoConsumers(ctx, stream, c.options.GroupName)
+		cs, err := cu.Result()
+		if err != nil {
+			c.Errors <- errors.New(fmt.Sprintf("failed to list consumers for stream: %q group: %q", stream, c.options.GroupName))
```
Comment (ndreas) [inline consumer.go]: `fmt.Errorf()` is better.

File: consumer.go [hunk 7]
```diff
@@ -218,9 +223,57 @@ func (c *Consumer) Run() {
...
+			}
+		}
+	}
+
+	// blocking, waiting for shutdown
 	c.wg.Wait()
 }
 
+func (c *Consumer) removeConsumerFromStream(stream string) {
+	fmt.Println("removeConsumerFromStream:" + stream)
```
Comment (ndreas) [inline consumer.go]: Debug output? Should this be a proper log instead?

File: consumer.go [hunk 8]
```diff
@@ -218,9 +223,57 @@ func (c *Consumer) Run() {
...
+	name := c.options.Name
+	rCmd := c.redis.XGroupDelConsumer(context.Background(), stream, c.options.GroupName, c.options.Name)
+	pendingMsgCount, err := rCmd.Result()
+	if err != nil {
+		c.Errors <- errors.New(fmt.Sprintf("failed to remove consumer: %q for stream: %q group: %q", name, stream, c.options.GroupName))
```
Comment (ndreas) [inline consumer.go]: dat `fmt.Errorf()`

File: consumer.go [hunk 9]
```diff
@@ -218,9 +223,57 @@ func (c *Consumer) Run() {
...
+	}
+	if pendingMsgCount > 0 {
+		c.Errors <- errors.New(fmt.Sprintf("remove consumer: %q with pending messages: %d for stream: %q group: %q", name, pendingMsgCount, stream, c.options.GroupName))
```
Comment (ndreas) [inline consumer.go]: very `fmt.Errorf()`

File: cmd/cleanup/cleanup.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+				// prevent ddos on redis
+				time.Sleep(10 * time.Millisecond)
+			} else {
+				fmt.Print("skipping\n")
+			}
+
+		}
+	}
+
+	fmt.Println("Bye!")
```
Comment (ndreas) [inline cmd/cleanup/cleanup.go:66]: :wave:

**Review by ndreas — CHANGES_REQUESTED** [review]
Comments and stuffs. The `fmt.Errorf()` comments I think should be addressed, otherwise, feel free to ignore.

File: cmd/cleanup/cleanup.go (hunk 3 above)
Comment (preciselytord) [reply cmd/cleanup/cleanup.go]: i should change the comment, but the time i 1 hour in millis * 24 = 1 day worth of idle

Comment (ndreas) [reply cmd/cleanup/cleanup.go]: I read it too fast, missed the hour :D

**Review by ndreas — APPROVED** [review]
:100:

**Pull request merged**